### PR TITLE
[2765] Add support for portal representation

### LINKED ISSUE #2765: Add support for a "Portal" representation
A portal can be created anywhere (on any semantic element in the project) and is just a kind of dashboard where the end-user can add other representations (of any kind) so that they can be all opened at the same time.

See https://github.com/eclipse-sirius/sirius-web/pull/2691 for the corresponding shape.

A more detailed ADR focusing on the technical aspects will be proposed soon.

### PR DESCRIPTION
**Pull request by pcdavid**

### CODE REVIEW

File: packages/portals/backend/pom.xml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ Contributors:
+     Obeo - initial API and implementation
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+
+	<groupId>org.eclipse.sirius</groupId>
+	<artifactId>sirius-components-portals-parent</artifactId>
+	<version>2023.10.9</version>
```
Comment (sbegaudeau) [inline packages/portals/backend/pom.xml]: Will have to be updated

File: packages/portals/backend/sirius-components-collaborative-portals/src/main/java/org/eclipse/sirius/components/collaborative/portals/PortalEventProcessor.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+    private final IEditingContext editingContext;
+
+    private final IRepresentationPersistenceService representationPersistenceService;
+
+    private final List<IPortalEventHandler> portalEventHandlers;
+
+    private final ISubscriptionManager subscriptionManager;
+
+    private final AtomicReference<Portal> currentPortal;
```
Comment (sbegaudeau) [inline packages/portals/backend/sirius-components-collaborative-portals/src/main/java/org/eclipse/sirius/components/collaborative/portals/PortalEventProcessor.java]: I don't think we need an atomic reference here, a regular field should work just fine but that's the same issue in other representations (and of course this would have no sense in being final)

File: packages/portals/frontend/sirius-components-portals/src/hooks/usePortalMutations.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        id
+      }
+      ... on ErrorPayload {
+        message
+      }
+    }
+  }
+`;
+
+export const usePortalMutations = (editingContextId: string, portalId: string): UsePortalMutationsValue => {
```
Comment (sbegaudeau) [inline packages/portals/frontend/sirius-components-portals/src/hooks/usePortalMutations.ts:92]: Add support for displaying errors with toast for all mutations

File: packages/portals/frontend/sirius-components-portals/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "$schema": "https://json.schemastore.org/tsconfig",
+  "extends": "@eclipse-sirius/sirius-components-tsconfig/react-library.json",
```
Comment (sbegaudeau) [inline packages/portals/frontend/sirius-components-portals/tsconfig.json]: Please start (or switch quickly) to the strict version for such as new project

File: packages/sirius-web/frontend/sirius-web/src/portals.css [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+
+.react-grid-item.react-grid-placeholder {
+  background: var(--blue-lagoon);
```
Comment (sbegaudeau) [inline packages/sirius-web/frontend/sirius-web/src/portals.css]: Use the proper color, blue-lagoon must have been deprecated years ago? It's mostly used by the soon to be deleted sporty code now.

Comment (sbegaudeau) [reply packages/sirius-web/frontend/sirius-web/src/portals.css]: Can't this file be located in sirius-components-portals frontend instead? Or could this be done another way but directly in sirius-components-portals to not have everybody do this in their application.

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
I have not executed the code but since I've seen it multiple times, I believe that it should work. Could you add a very basic Cypress test creating an empty portal and verifying that it is displayed too in order to ensure that we don't break everything without noticing it, thanks.

File: packages/portals/backend/pom.xml (hunk 1 above)
Comment (pcdavid) [reply packages/portals/backend/pom.xml]: Done.

File: packages/portals/backend/sirius-components-collaborative-portals/src/main/java/org/eclipse/sirius/components/collaborative/portals/PortalEventProcessor.java (hunk 2 above)
Comment (pcdavid) [reply packages/portals/backend/sirius-components-collaborative-portals/src/main/java/org/eclipse/sirius/components/collaborative/portals/PortalEventProcessor.java]: Done (only for this processor).

File: packages/portals/frontend/sirius-components-portals/tsconfig.json (hunk 4 above)
Comment (pcdavid) [reply packages/portals/frontend/sirius-components-portals/tsconfig.json]: Done. And fixed the few typing issues it revealed.

File: packages/portals/frontend/sirius-components-portals/src/hooks/usePortalMutations.ts (hunk 3 above)
Comment (pcdavid) [reply packages/portals/frontend/sirius-components-portals/src/hooks/usePortalMutations.ts:92]: Done. I've attempted a semi-generic (at least locally) hook for reporting errors from mutations, comments welcome.

File: packages/sirius-web/frontend/sirius-web/src/portals.css (hunk 5 above)
Comment (pcdavid) [reply packages/sirius-web/frontend/sirius-web/src/portals.css]: I tried that, but couldn't make it work.

* Put the style in `packages/portals/frontend/sirius-components-portals/src/representations/portal.css`.
* Added `import './portal.css';` in `PortalRepresentation.tsx`.
* Added `css()` from `rollup-plugin-import-css` in the Vite config (and added `rollup-plugin-import-css` in `devDependencies`).

My custom color is not picked up. Do you see what I'm missing?

Comment (gcoutable) [reply packages/sirius-web/frontend/sirius-web/src/portals.css]: ~`import './portal.css';` can be done in `packages/sirius-web/frontend/sirius-web/index.tsx` like others.~
~I don't have any other solution in mind.~
That was the initial location, don't mind the answer.

**Comment by pcdavid** [comment]
Note: the only test failing is in `project/diagrams/diagram-contextual-palette.cy.js`. It's completely unrelated to this PR, and unfortunately known to be unreliable.

**Review by sbegaudeau — APPROVED** [review]
I'lLet's approve this to be able to merge it. @pcdavid feel free to do it, since I have a couple of other things to review you may end up doing it faster than me, in case it needs to be rebased for example.
